packages/destination-actions/src/destinations/aws-s3/uploadCsv/s3.ts

```
@@ -0,0 +1,136 @@
+  const stsClient = new STSClient({ region })
+  const command = new AssumeRoleCommand({
+    RoleArn: roleArn,
+    RoleSessionName: roleSessionName
+  })
+
+  const response = await stsClient.send(command)
+  if (!response.Credentials) {
+    throw new Error('Failed to assume role and get temporary credentials')
+  }
+
+  return {
+    accessKeyId: response.Credentials.AccessKeyId!,
+    secretAccessKey: response.Credentials.SecretAccessKey!,
+    sessionToken: response.Credentials.SessionToken!
+  }
+}
+
+function validateS3(payload: Payload) {
+  if (!payload.iam_role_arn) {
+    throw new InvalidAuthenticationError('Selected S3 upload mode, but missing IAM Role ARN')
+  }
+  if (!payload.s3_aws_bucket_name) {
+    throw new InvalidAuthenticationError('Selected S3 upload mode, but missing AWS S3 bucket name')
+  }
+
+  if (payload.s3_aws_folder_name) {
+    if (!payload.s3_aws_folder_name.endsWith('/')) {
+      throw new InvalidAuthenticationError('Selected S3 upload mode, but the folder name must end with "/"')
+    }
+  }
+
+  if (!payload.s3_aws_region) {
+    throw new InvalidAuthenticationError('Selected S3 upload mode, but missing AWS Region')
+  }
+}
+
+// Function to upload a CSV content to S3
+const uploadCSV = async (
+  credentials: Credentials,
+  fileContent: Buffer | string,
+  bucketName: string,
+  folderName: string,
+  key: string,
+  region: string
+) => {
+  const s3Client = new S3Client({
+    region,
+    credentials: {
+      accessKeyId: credentials.accessKeyId,
+      secretAccessKey: credentials.secretAccessKey,
+      sessionToken: credentials.sessionToken
+    }
+  })
+  const objectKey = folderName ? `${folderName}${key}` : key
+  const uploadParams = {
+    Bucket: bucketName,
+    Key: objectKey,
+    Body: fileContent,
+    ContentType: 'text/csv'
+  }
+
+  try {
+    const data = await s3Client.send(new PutObjectCommand(uploadParams))
+    console.log('Upload Success', data)
+  } catch (err) {
+    console.error('Error', err)
```

We don't log errors like this in destinations. Either do a non-retryable error code (400) or let the method fail

---

Added in correct error throwing